SBT : Version 0.0.1

ITStar10 opened the pull request.

ITStar10 committed:
Working : Extends ERC721URIStorageUpgradeable

To-do :
	Remove dpendency to ERC721URIStorageUpgradeable
	Extends ERC721MetadataUpgradeable

ITStar10 committed:
Working : Extends ERC721URIStorageUpgradeable

To-do :
	Remove dpendency to ERC721URIStorageUpgradeable
	Extends ERC721MetadataUpgradeable

ITStar10 committed:
Test code not done

Updated package.json of following contracts to share code:
	VDA-Name-Registry
	VDA-Verification-Base

Update one function in VDA-Verification-Base

Updated ReadMe

ITStar10 committed:
Test code not done

Updated package.json of following contracts to share code:
	VDA-Name-Registry
	VDA-Verification-Base

Update one function in VDA-Verification-Base

Updated ReadMe

ITStar10 committed:
Initial version : 0.0.1

Test done by test code

ITStar10 committed:
Initial version : 0.0.1

Test done by test code

ITStar10 committed:
V0.0.1 - Initial version (#61)

* V0.0.1 - Initial version

Implemented Verida VIP-2 and tested

* Updates for Controller & Revoke

Added:
	Controller management
	Revoke

Updated:
	Test code

* Updates on review

Reject registereing for revoked DIDs.

tahpot commented on VDA-SBT/contracts/ISoulboundNFT.sol:
```diff
@@ -0,0 +1,125 @@
+     * @dev Only owner can do this
+     * @param account Company account to be added
+     */
+    function addCompanyAccount(address account) external;
```
Rename to "addTrustedAddress()"

tahpot commented on VDA-SBT/contracts/ISoulboundNFT.sol:
```diff
@@ -0,0 +1,125 @@
+     * @dev Only owner can do this
+     * @param account Company account to be removed
+     */
+    function removeCompanyAccount(address account) external;
```
Rename to "removeTrustedAddress()"

tahpot commented on VDA-SBT/contracts/ISoulboundNFT.sol:
```diff
@@ -0,0 +1,125 @@
+     * @dev Only owner can see this
+     * @return address[] list of company accounts
+     */
+    function listCompanyAccounts() external view returns(address[] memory);
```
Rename to "getTrustedAddresses()"

tahpot commented on VDA-SBT/contracts/ISoulboundNFT.sol:
```diff
@@ -0,0 +1,125 @@
+     * @dev Only the owner can add
+     * @param sbtType new type to be added
+     */
+    function addSBTType(string calldata sbtType) external;
```
Could we possibly **not** have a list of SBT types and allow the "sbtType" when claiming to be any valid string?

This allows us to easily support any "sbtType" which is going to be very helpful.

ITStar10 replied:
So any valid string can be SBT type, correct?

tahpot replied:
Yes, correct.

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+     * @notice Claimed SBT info by SBT type
+     * @dev mapping of SBTType => (user => tokenId)
+     */
+    mapping(string => mapping(address => uint)) private _sbtInfo;
```
Can we change this to be user => (sbttype => tokenId)?

I think that will be more efficient, especially if we don't have a pre-coded list of sbtTypes

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+
+import "hardhat/console.sol";
+
+contract SoulboundNFT is VDAVerificationContract, 
```
This contract needs to throw for all transfer related functions. See EIP-5192 specification: In this case, all EIP-721 functions of the contract that transfer the token from one account to another must throw.

>A token with a uint256 tokenId may be bound to a receiving account with function locked(...) returning true. In this case, all EIP-721 functions of the contract that transfer the token from one account to another must throw.

ITStar10 replied:
Already implemented.
_beforeTokenTransfer() function is called on every transfer related function. There it throws an error.

tahpot replied:
Got it.

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+        override(IERC721MetadataUpgradeable, ERC721Upgradeable)
+        returns (string memory)
+    {
+        return ERC721Upgradeable.tokenURI(tokenId);
```
Need to return the tokenUri saved when "claim" is called.

ITStar10 replied:
Now, in the claimSBT() function, it returns  tokenID. You want to return tokenURI instead?

tahpot replied:
The current code is good. Not sure if you changed something or I misread it the first time :)

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+        string calldata sbtType,
+        string calldata uniqueId,
+        bytes calldata signature,
+        bytes calldata proof
```
We need to add a new field "tokenUri" that is saved with each token.

ITStar10 replied:
You mean, we sould set tokenURI on claim?

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+                "-",
+                didNonce
+            );
+            verifyRequest(did, params, signature, proof);
```
@todo This verification code will need to use the Verida Verification Base Contract when it's ready.

ITStar10 replied:
SBT contract can't use the VerificationBaseContract's verifyRequest() function.
Because in the VerificationBaseContract, verifyRequest() function verify the signer is the `did` in the input parameters.
But in the SBT contract, we should iterate all the company accounts to verify the signature is signed by one of registered company accounts.

ITStar10 replied:
So, in the SouboundNFT contract, I overrided the `verifyRequest(...)` function.

tahpot replied:
>Because in the VerificationBaseContract, verifyRequest() function verify the signer is the did in the input parameters.
But in the SBT contract, we should iterate all the company accounts to verify the signature is signed by one of registered company accounts.

The `company accounts` are actually Verida DID's not Polygon addresses.

ITStar10 replied:
Please check the verifyRequest() function in the VerificationBaseContract. 
![image](https://user-images.githubusercontent.com/15656252/203179005-b6c243c0-d137-42bb-950f-276dde759278.png)
Once the recovered address doesn't equal to the inputed `did`, it will reject the entire transaction.
In our case, we should iterate through all our company address, and check one of them is matched.
So, we can't use the original verifyRequest function.

tahpot replied:
Yes. I think we should change verifyRequest to provide an array if `dids`:

`verifyRequest(address[] dids, ...)`

and then 

`require(proofSigner in dids)`

(pseudocode)

ITStar10 replied:
Means we should update verifyRequest() function in the VerificationBase contract?

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -0,0 +1,316 @@
+    /**
+     * @dev See {IsoulboundNFT}
+     */
+    function claimSBT(
```
The caller of this method may be the metadata server, so we need to support the caller specifying the address that should receive the minted NFT

tahpot reviewed (changes requested):
This is a great start. See my comments in the code.

ITStar10 committed:
Name-Registry : Minor updates (#64)

* Name-Registry : Minor updates

Update
	Renamed getNonce() function to nonce()
	Deployment script to save the addresses

* Minor update on  hardhat configuratino

ITStar10 committed:
VDA-DID-Registry Update lookup() function (#66)

- Update the return value of lookup() function
- Update test codes
- Deploy on Mumbai net

ITStar10 committed:
Minor updates : Function name & events name updated

ITStar10 committed:
VDA-SBT : Updated on Review

Updated:
	claimSBT() function updated
	Test code updated

VDA-Verification-Base contract:
	Update verifyRequest() function

ITStar10 committed:
Merge branch 'feature/57-minimal-SBT' of github.com:verida/blockchain-contracts into feature/57-minimal-SBT

ITStar10 committed:
Minor updates on review

Removed
	allowedSBTTypes() function
	unused events
